### PR TITLE
📚 docs: add research report for runs query implementation

### LINKED ISSUE #299: 298.1-research Research langsmith-sdk runs query precedent
## Overview

Research sub-issue for #298 (ls-runs-query milestone) to investigate the langsmith-sdk Python implementation and establish recommendations for the Rust implementation.

## Parent Issue

Part of #298 - ls-runs-query milestone

## Scope

1. **Analyze langsmith-sdk Python client** (`/workspace/reference/repo/langchain-ai/langsmith-sdk/code/python/langsmith/client.py`)
   - Document the `list_runs` method signature and parameters
   - Understand filter query language implementation
   - Identify pagination patterns
   - Note tree_filter support

2. **Review async client** (`async_client.py`)
   - Compare sync vs async implementations
   - Identify patterns useful for Rust

3. **Examine schemas** (`schemas.py`)
   - Document Run schema fields
   - Identify which fields are queryable/filterable

4. **Review tests** (`tests/integration_tests/test_runs.py`)
   - Extract usage patterns and examples
   - Identify edge cases handled

5. **Check REST API patterns**
   - Document `/runs/query` endpoint usage
   - Identify request/response shapes

## Deliverable

Create a research report at `reference/research/298-ls-runs-query-precedent.md` documenting:
- Recommended Rust implementation approach
- API client patterns to follow
- Filter builder design recommendations
- Pagination strategy
- Testing approach

## Acceptance Criteria

- [ ] Research report created and committed
- [ ] Report covers all scope items
- [ ] Recommendations are actionable for Rust implementation

### PR DESCRIPTION
**Pull request by codekiln**
## Summary

Closes #299

This PR adds a comprehensive research report analyzing the LangSmith Python SDK's `list_runs` implementation to establish recommendations for implementing `langstar runs query` in Rust.

### Changes

- **New research report**: `reference/research/298-ls-runs-query-precedent.md`
  - Analyzes Python SDK's `list_runs` method signature and parameters
  - Documents the LangSmith filter query language (function-style syntax)
  - Details cursor-based pagination implementation
  - Provides recommended Rust struct definitions and method signatures
  - Includes implementation phases and testing strategy

- **Updated workflow documentation**: `docs/dev/github-workflow.md`
  - Added "Choosing PR Target: Main vs. Hierarchical Merging" section
  - Clarifies when to use direct-to-main PRs vs hierarchical merging
  - Provides decision flowchart for choosing PR strategy
  - Documents that hierarchical merging should only be used for breaking changes

### Key Findings

1. **Endpoint**: Use `POST /runs/query` (not GET)
2. **Pagination**: Cursor-based with `cursors.next` token
3. **Filter syntax**: Function-style (`eq(field, value)`, `and()`, `or()`, etc.)
4. **Field mapping**: `session` = project IDs in request body

## Test Plan

- [x] Research report is comprehensive and actionable
- [x] Workflow documentation update is clear and provides good guidance
- [x] Report includes specific code recommendations for Rust implementation

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

Co-Authored-By: Claude <noreply@anthropic.com>

### CODE REVIEW

**Comment by codekiln** [comment]
Addressed review comments in commit 1fa444d:

- **to_string shadowing**: Renamed `to_string` to `to_filter_string()` to avoid shadowing the standard library's `ToString` trait method
- **Missing format() method**: Added the `format()` method implementation to `FilterValue` enum for proper value serialization in filter expressions

File: reference/research/298-ls-runs-query-precedent.md [hunk 1]
```diff
@@ -0,0 +1,573 @@
...
+        match self {
+            FilterOp::Eq(field, val) => format!("eq({}, {})", field, val.format()),
+            FilterOp::And(ops) => {
+                let parts: Vec<_> = ops.iter().map(|o| o.to_string()).collect();
+                format!("and({})", parts.join(", "))
+            }
+            // ... other operators
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (codekiln) [reply reference/research/298-ls-runs-query-precedent.md]: Fixed in 1fa444d - renamed `to_string` to `to_filter_string()` to avoid shadowing the standard library's ToString trait method.
